### PR TITLE
Parametric Block Non-linear form with examples

### PR DESCRIPTION
**Pull request by bslazarov**
The new code provides parametric block non-linear form necessary for shortening the realization of optimization problems. The usage is exemplified with two examples in the mini-app folder mtop. Additional examples including complete topology optimization mini-app are under preparation. The current version demonstrates only the evaluation of gradients.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2079](https://github.com/mfem/mfem/pull/2079) | @bslazarov | @tzanio | @jamiebramwell + @jandrej + @v-dobrev | 03/04/21 | 04/29/21 | 05/10/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: miniapps/mtop/mtop_integrators.hpp
Comment (jamiebramwell) [inline miniapps/mtop/mtop_integrators.hpp:21]: As this will be the only example of a working `ParametricBNLFormIntegrator`, I think this auxilliary `BaseQFunction` class unnecessarily complicates things.

File: miniapps/mtop/mtop_integrators.hpp
Comment (jamiebramwell) [inline miniapps/mtop/mtop_integrators.hpp:55]: This class needs comments

File: fem/prmnonlinearform.hpp
Comment (bslazarov) [reply fem/prmnonlinearform.hpp:117]: cleaned

File: fem/pprmnonlinearform.hpp
Comment (bslazarov) [reply fem/pprmnonlinearform.hpp:45]: renamed

File: fem/pprmnonlinearform.hpp
Comment (bslazarov) [reply fem/pprmnonlinearform.hpp:61]: Extended

File: fem/prmnonlinearform.hpp
Comment (bslazarov) [reply fem/prmnonlinearform.hpp:33]: all names are renamed to Param*, *param and param*

File: fem/prmnonlinearform.hpp
Comment (bslazarov) [reply fem/prmnonlinearform.hpp:1]: done

File: fem/pprmnonlinearform.hpp
Comment (bslazarov) [reply fem/pprmnonlinearform.hpp:66]: added

File: fem/prmnonlinearform.cpp
Comment (bslazarov) [reply fem/prmnonlinearform.cpp:35]: The new description is there.

File: miniapps/mtop/mtop_integrators.hpp
Comment (bslazarov) [reply miniapps/mtop/mtop_integrators.hpp:74]: I would prefer to leave it like this - Most probably the implementation will be modified in the near future when the new examples for topology optimization are added in the code.

File: miniapps/mtop/mtop_integrators.cpp
Comment (bslazarov) [reply miniapps/mtop/mtop_integrators.cpp:150]: Removed - thanks!

File: miniapps/mtop/mtop_integrators.hpp
Comment (bslazarov) [reply miniapps/mtop/mtop_integrators.hpp:21]: It will be used for the topology optimization miniapp. Otherwise one would have to implement a new version and may be new interface for every new physics.

File: miniapps/mtop/mtop_integrators.hpp
Comment (bslazarov) [reply miniapps/mtop/mtop_integrators.hpp:55]: I have improved the comments - Additional comments will be added with the topopt miniapp

**Comment by jandrej** [comment]
I'd like to see a consistent comment and formatting style. Some examples of individual comments throughout
```
//set the matrix B
//returns the derivative, with respect to the density, of the product
//Provides implementation of an integrator for linear diffusion
// Computes an example of nonlinear objective
//          Define a parallel mesh by a partitioning of the serial mesh. Refine
```
It would be good to unify and settle for a style of indentation and capitalization that is used in the code. Most class members and some classes miss doxygen style comment blocks.

I would prefer to have spacing on assignments, but if this passed style checking, I have no power here.
Most variable names are still cryptic. I can't make sense of `mfem::BlockVector grdbv` unless I read the whole code. This is not desirable.

The naming of the main addition `ParametricBNLFormIntegrator` on the other hand is very verbose. I have no better idea right now in terms of naming.

I know that there is a specific purpose for this implementation. And this is also the point why I don't see this as a good fit for the library. It adds a convenience layer by duplicating functionality of the BlockNonlinearForm and adding parametric fields.

Don't get me wrong, the functionality and implementation is perfectly fine. In my point of view I would rather see a, possibly breaking, interface change in BlockNonlinearForm and putting this functionality in the existing implementation. The addition right now would lead to a maintainability burden that would include 4 classes in total (BlockNLF + ParametricBNLF + Par for each).

**Comment by bslazarov** [comment]
> I'd like to see a consistent comment and formatting style. Some examples of individual comments throughout
> 
> ```
> //set the matrix B
> //returns the derivative, with respect to the density, of the product
> //Provides implementation of an integrator for linear diffusion
> // Computes an example of nonlinear objective
> //          Define a parallel mesh by a partitioning of the serial mesh. Refine
> ```
> 
> It would be good to unify and settle for a style of indentation and capitalization that is used in the code. Most class members and some classes miss doxygen style comment blocks.

The comments are made consistent in the revised code.

**Comment by bslazarov** [comment]
> Most variable names are still cryptic. I can't make sense of `mfem::BlockVector grdbv` unless I read the whole code. This is not desirable.

Comments have been added for all variables. Longer names will not improve the readability. The current examples are intended to be read step-by-step. The current naming is consistent all block vectors end with bv and the first three letters  stand for gradient, adjoint and primal. 


> 
> The naming of the main addition `ParametricBNLFormIntegrator` on the other hand is very verbose. I have no better idea right now in terms of naming.

The name came after a discussion with Jamie. If you have shorter or more explanatory name, please, let me know. 
> 
> I know that there is a specific purpose for this implementation. And this is also the point why I don't see this as a good fit for the library. It adds a convenience layer by duplicating functionality of the BlockNonlinearForm and adding parametric fields.
> 
> Don't get me wrong, the functionality and implementation is perfectly fine. In my point of view I would rather see a, possibly breaking, interface change in BlockNonlinearForm and putting this functionality in the existing implementation. The addition right now would lead to a maintainability burden that would include 4 classes in total (BlockNLF + ParametricBNLF + Par for each).

Adding the parametric form on top of the non-linear block form would require rewriting a quite significant part of MFEM. The parametric fields require additional members in the signatures of the methods in the integrators and the forms. These are irrelevant for non-parametric cases and, apart from the huge amount of work, will decrease the readability of the code. As the BNL form is intended to be used mainly for optimization, one way to avoid the propagation of dependencies in the core MFEM library is to move the form and the integrators in the miniapp directory. Let me know if such a solution will be better in this case.

**Comment by jandrej** [comment]
If the implementation is moved in a miniapp directory I don’t see a problem with it.

I don’t know if thats feasible for you. But I don’t have the last word and I’m certainly not forcing my opinion or “deciding”. It might be good invovle @tzanio  or @v-dobrev to take a look.

File: fem/nonlininteg.hpp
Comment (tzanio) [inline fem/nonlininteg.hpp:155]: Please revert the changes in this file

File: miniapps/mtop/mtop_integrators.cpp
Comment (tzanio) [inline miniapps/mtop/mtop_integrators.cpp:1]: Add the Copyright header

**Comment by tzanio** [comment]
I am OK with the current structure. Please also

- [ ] Fix Travis
- [ ] Update `CHANGELOG`

**Review by jandrej — APPROVED** [review]

File: CHANGELOG
Comment (jandrej) [inline CHANGELOG:112]: s/min-app/mini-app

File: CHANGELOG
Comment (jandrej) [inline CHANGELOG:113]: remove extra space

**Comment by v-dobrev** [comment]
> If the implementation is moved in a miniapp directory I don’t see a problem with it.
> 
> I don’t know if thats feasible for you. But I don’t have the last word and I’m certainly not forcing my opinion or “deciding”. It might be good invovle @tzanio or @v-dobrev to take a look.

To me, the parametric block nonlinear form makes sense to be part of the core mfem classes because it could be useful beyond the mtop miniapp, although I don't have specific use cases in mind. However, for now, I'm not strongly opposed to keep it as part of the miniapp.

@jamiebramwell, what do you think?

**Comment by jamiebramwell** [comment]
> 
> 
> > If the implementation is moved in a miniapp directory I don’t see a problem with it.
> > I don’t know if thats feasible for you. But I don’t have the last word and I’m certainly not forcing my opinion or “deciding”. It might be good invovle @tzanio or @v-dobrev to take a look.
> 
> To me, the parametric block nonlinear form makes sense to be part of the core mfem classes because it could be useful beyond the mtop miniapp, although I don't have specific use cases in mind. However, for now, I'm not strongly opposed to keep it as part of the miniapp.
> 
> @jamiebramwell, what do you think?

Sounds good to me, I just wanted to get your sign-off before adding another assembly routine that needs to be maintained.

**Comment by tzanio** [comment]
My preference actually is to keep things in the subdirectory for now, and move to the core library if/when necessary.

**Comment by v-dobrev** [comment]
> Sounds good to me, I just wanted to get your sign-off before adding another assembly routine that needs to be maintained.

@jamiebramwell, I don't quite understand: which version sounds good to you? (1) Keep the parametric block nonlinear form in the miniapp or (2) make it part of the core mfem library?

**Comment by jamiebramwell** [comment]
> 
> 
> > Sounds good to me, I just wanted to get your sign-off before adding another assembly routine that needs to be maintained.
> 
> @jamiebramwell, I don't quite understand: which version sounds good to you? (1) Keep the parametric block nonlinear form in the miniapp or (2) make it part of the core mfem library?

I was saying moving it into the core library, but I think my opinion should get overruled by you and @tzanio .

File: miniapps/mtop/mtop_integrators.hpp
Comment (jamiebramwell) [reply miniapps/mtop/mtop_integrators.hpp:94]: I suppose we will have to agree to disagree on this point. I would very much prefer more verbose names as it requires less "context switching" when reading code, but if @tzanio and @v-dobrev are OK with this, then I will not hold up this PR.

File: miniapps/mtop/mtop_integrators.cpp
Comment (jamiebramwell) [inline miniapps/mtop/mtop_integrators.cpp:115]: I'm a bit confused here. Will `ir` ever not be `nullptr`?

File: miniapps/mtop/mtop_integrators.cpp
Comment (jamiebramwell) [inline miniapps/mtop/mtop_integrators.cpp:187]: Same issue here. Why does this check exist?

**Review by jamiebramwell — APPROVED** [review]

**Comment by jamiebramwell** [comment]
One more TODO item:

- [ ] Add a documentation PR in the [web repo](https://github.com/mfem/web).

File: miniapps/mtop/mtop_integrators.hpp
Comment (v-dobrev) [reply miniapps/mtop/mtop_integrators.hpp:94]: Is this about `dd` and `uu`? I think adding doxygen for the parameters (in the base class) will help more.

Comment (tzanio) [reply miniapps/mtop/mtop_integrators.hpp:94]: I think both will help, but maybe it is not urgent if the code will change :)

**Comment by tzanio** [comment]
@bslazarov, is this ready for merging in `next`?

File: miniapps/mtop/mtop_integrators.cpp
Comment (bslazarov) [reply miniapps/mtop/mtop_integrators.cpp:115]: Thanks - that is left over from the old code. Corrected in the new version.

File: miniapps/mtop/mtop_integrators.cpp
Comment (bslazarov) [reply miniapps/mtop/mtop_integrators.cpp:187]: corrected

**Comment by bslazarov** [comment]
> One more TODO item:
> 
>     * [ ]  Add a documentation PR in the [web repo](https://github.com/mfem/web).

I will wait a bit for the full topopt app before adding the documentation

**Comment by bslazarov** [comment]
> @bslazarov, is this ready for merging in `next`?

yes, please, go ahead

**Comment by jamiebramwell** [comment]
> 
> 
> > One more TODO item:
> > ```
> > * [ ]  Add a documentation PR in the [web repo](https://github.com/mfem/web).
> > ```
> 
> I will wait a bit for the full topopt app before adding the documentation

If this is included in v4.3, I really think there should be at least some minimal documentation. However, I will defer to the opinion of @tzanio and @v-dobrev .

**Comment by bslazarov** [comment]
> > > One more TODO item:
> > > ```
> > > * [ ]  Add a documentation PR in the [web repo](https://github.com/mfem/web).
> > > ```
> > 
> > 
> > I will wait a bit for the full topopt app before adding the documentation
> 
> If this is included in v4.3, I really think there should be at least some minimal documentation. However, I will defer to the opinion of @tzanio and @v-dobrev .

Ok, then I will add documentation for gradient evaluation :-)

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
See also https://github.com/mfem/web/issues/109

**Comment by tzanio** [comment]
LGTM

**Pull request merged**
